### PR TITLE
Targets, Distributed, and RL API updates.

### PR DESCRIPTION
**Pull request by wenkesj**
WIP to support distributed RL algorithms, including V-trace and R2D2 -- (don't merge yet)

Making some updates to the RL API and losses.

### CODE REVIEW

**Comment by jimfleming** [comment]
Can you base the PR on the `rainbow` branch?

File: pyoneer/rl/rollouts/gym_ops.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+
+from gym import spaces
+
+import tensorflow as tf
+
+from pyoneer.debugging import debugging_ops
+from pyoneer.rl.wrappers import batch_impl
+
+
+def space_to_spec(space):
```
Comment (jimfleming) [inline pyoneer/rl/rollouts/gym_ops.py]: This is cool!

File: pyoneer/rl/rollouts/gym_ops.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+                             tf.dtypes.as_dtype(space.dtype))
+    elif isinstance(space, spaces.Tuple) or isinstance(space, spaces.Dict):
+        return tf.nest.map_structure(space_to_spec, space.spaces)
+    raise TypeError('`space` not supported: {}'.format(type(space)))
+
+
+class Transition(collections.namedtuple(
+        'Transition', ['state', 'reward', 'next_state', 'terminal', 'weight'])):
+
+    __slots__ = ()
```
Comment (jimfleming) [inline pyoneer/rl/rollouts/gym_ops.py]: This seems like a micro-optimization and is sufficiently uncommon to require a comment at minimum. :)

File: pyoneer/rl/rollouts/gym_ops.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+
+class Env(object):
```
Comment (jimfleming) [inline pyoneer/rl/rollouts/gym_ops.py]: `TensorFlowEnv`?

File: pyoneer/rl/rollouts/unroll_ops.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+import tensorflow as tf
+
+
+class Unroll(collections.namedtuple(
+        'Unroll', ['outputs', 'initializer', 'initial_time_step'])):
+    pass
+
+
+def n_step_unroll(n_step,
```
Comment (jimfleming) [inline pyoneer/rl/rollouts/unroll_ops.py]: 👍 I like the option of having a function or class.

File: pyoneer/rl/rollouts/unroll_ops.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+                  initializer,
+                  agent_step_fn,
+                  env_step_fn,
+                  initial_time_step=0):
+    """Unroll interactions for n-steps.
+
+    Args:
+        n_step: Tensor of shape [].
```
Comment (jimfleming) [inline pyoneer/rl/rollouts/unroll_ops.py]: Integer, right?

File: pyoneer/rl/rollouts/unroll_ops.py [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+        initializer: Tuple containing initial values:
+            (agent_output, env_output)
+        agent_step_fn: Agent step function:
+            fn(env_output, agent_output, time_step)
+        env_step_fn: Environment step function:
+            fn(agent_output, env_output, time_step)
+        initial_time_step: The initial step value.
```
Comment (jimfleming) [inline pyoneer/rl/rollouts/unroll_ops.py]: What's the use case of having an argument for this?

File: pyoneer/rl/rollouts/unroll_ops.py [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+
+    Returns:
+        Tuple containing:
+            - Tensors of shape [Time x ...] with same structure as
```
Comment (jimfleming) [inline pyoneer/rl/rollouts/unroll_ops.py]: Does this default to time-major when using `Batch`?

File: pyoneer/rl/rollouts/unroll_ops.py [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+    def __init__(self, env, agent, n_step, time_major=False):
+        """Creates a new Rollout.
+
+        Nest-compatible API for arbitrarily nested transitions.
+
+        Args:
+            env: The Env interface.
+            agent: The Agent interface.
+            n_step: The number of steps to rollout.
+            time_major: Time major outputs.
```
Comment (jimfleming) [inline pyoneer/rl/rollouts/unroll_ops.py]: What's the use case of having one or the other? I feel like we should just pick one and only use that.

File: pyoneer/rl/rollouts/unroll_ops.py [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+        """
+        self._env = env
+        self._agent = agent
+        self._n_step = n_step
+        self._time_major = time_major
+
+    @tf.function
+    def __call__(self, initializer=None, initial_time_step=0):
```
Comment (jimfleming) [inline pyoneer/rl/rollouts/unroll_ops.py]: I'm not sure I understand the initializer. What does it do?

File: pyoneer/rl/wrappers/action_impl.py [hunk 10]
```diff
@@ -8,7 +8,7 @@
 from pyoneer.math import math_ops
 
 
-class ActionScale(gym.Wrapper):
+class ActionScale(gym.ActionWrapper):
```
Comment (jimfleming) [inline pyoneer/rl/wrappers/action_impl.py]: Nifty!

File: pyoneer/debugging/debugging_ops.py [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+
+    def __enter__(self):
+        self.start()
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.stop()
+
+
+def mock_spec(batch_shapes, specs, initializers=tf.zeros):
```
[the remark this answers is not in the document]
Comment (wenkesj) [reply pyoneer/debugging/debugging_ops.py]: This is intended for debugging and making initialization flexible with `tf.nest`. The idea is to not pass a constant around, instead pass a `tf.TensorSpec` that bears the spec for a tensor to be created with arbitrary shapes.

File: pyoneer/distributed/distributed_ops.py [hunk 12]
```diff
@@ -0,0 +1,318 @@
...
+        return tf.io.serialize_tensor(tensor)
+
+    def _decode(self, msg, dtype):
+        return tf.io.parse_tensor(msg, dtype)
+
+    def encode(self, item):
+        """Encode a structure into a packed message.
+
+        Args:
+            item: A structure with the same structure as dtypes.
```
[the remark this answers is not in the document]
Comment (wenkesj) [reply pyoneer/distributed/distributed_ops.py]: I need to work on the docs for this.

File: pyoneer/distributed/distributed_ops.py [hunk 13]
```diff
@@ -0,0 +1,318 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import tensorflow as tf
+
+
+class TensorCodec(object):
```
[the remark this answers is not in the document]
Comment (wenkesj) [reply pyoneer/distributed/distributed_ops.py]: This is a base class for encoding (and decoding) nested structures full of tensor-like objects to (and from) byte (protobuf) strings. It's intended use is to make passing nested structures through inter-process/remote-process fast (and possible w/ tools like redis).

File: pyoneer/rl/rollouts/unroll_ops.py (hunk 6 above)
Comment (wenkesj) [reply pyoneer/rl/rollouts/unroll_ops.py]: This allows you to track meta-data for rollouts seen from the agent/environment point-of-view in a stateful way. So if you wanted to resume a rollout where you left off, you would pass the final time step of the previous rollout as the initial time step of this rollout.

File: pyoneer/rl/rollouts/unroll_ops.py (hunk 9 above)
Comment (wenkesj) [reply pyoneer/rl/rollouts/unroll_ops.py]: This allows the rollout to be stateful, similar to the Keras RNN or RNNCell api, where you can pass arbitrary initial_states to the model. The initializer argument is a (optional) replacement for env.reset and agent.reset calls, so this skips over them and passes this in place of those return values.

File: Pipfile [hunk 14]
```diff
@@ -5,8 +5,8 @@ name = "pypi"
 
 [packages]
 gym = "*"
-tensorflow = "==2.0.0-beta1"
-tfp-nightly = "*"
+tensorflow = "==2.0.0"
```
Comment (djsaunde) [inline Pipfile]: :clap:

**Comment by djsaunde** [comment]
@wenkesj Don't forget to install the black pre-commit hook! This will run the black formatter when you run a `git commit` command.

**Pull request closed without merging**